aten/src/ATen/native/quantized/cpu/qmul.cpp

```diff
@@ -0,0 +1,91 @@
+#include <ATen/ATen.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/cpu/Loops.h>
+#include <ATen/quantized/Quantizer.h>
+
+#include <algorithm>
+
+namespace at {
+namespace native {
+namespace {
+
+inline void check_inputs(const Tensor& qa, const Tensor& qb) {
+  TORCH_CHECK(qa.qscheme() == kPerTensorAffine ||
+              qa.qscheme() == kPerTensorSymmetric,
+              "Only per tensor quantization is suuported in Mul.");
+  TORCH_CHECK(qa.qscheme() == qb.qscheme(),
+              "Both inputs to Mul must have the same quantization shceme.");
+  TORCH_CHECK(qa.numel() == qb.numel(),
+              "Mul operands must be the same size!");
+  TORCH_CHECK(qa.scalar_type() == qb.scalar_type(),
+              "Mul operands should have same data type.");
+}
+
+// Note: out is assumed to be the same size as self and other.
+// Note: Multiplication is only supported when self, other, out are of the same
+//       dtype.
+template <bool ReLUFused = false>
+Tensor _mul_out(Tensor& out, const Tensor& self, const Tensor& other) {
+  int64_t zero_point = out.q_zero_point();
+  double scale = out.q_scale();
+  int64_t self_zero_point = self.q_zero_point();
+  double self_scale = self.q_scale();
+  int64_t other_zero_point = other.q_zero_point();
+  double other_scale = other.q_scale();
+
+  auto iter = TensorIterator::binary_op(out, self, other);
+  AT_DISPATCH_QINT_TYPES(out.scalar_type(), "qmul", [&]() {
+    cpu_kernel(iter, [&](scalar_t a, scalar_t b) -> scalar_t {
+      const auto da = at::dequantize_val(self_scale, self_zero_point, a);
```

Comment:
don't we have a requantize_val API?

Reply:
I was afraid of the error contribution. I will try it out!

Reply:
Actually, I don't think that will help at all. The problem is that the multiplication doen't really work well with quantized types. My rationale is that if we requantize the values, we still need to compute `qc = round(a*b/s + z)`, where a, b are floating point multiplicands, and s, z are the output params. We are given `qa = round(a/s + z), qb = round(b/s + z)`, which are the requantized values. We are given `qa, qb` and need to compute `qc` I don't see a way on how to do it with out dequantizing them.

Reply:
OK, also noticed you are using const here, Tensor is a pointer so const is not necessary.

Reply:
I think this is a `scalar_t` value, not tensor